### PR TITLE
Remove hack skipping comparison of algebraic universes in subtyping.

### LINKED ISSUE #7695: Proof of False through modules and algebraic universes
```coq
Require Import Hurkens.

Universes i j k.
Module Type T.
  Parameter T1 : Type@{i+1}.
  Parameter e : Type@{j} = T1 : Type@{k}.
End T.

Module M.
  Definition T1 := Type@{j}.
  Definition e : Type@{j} = T1 : Type@{k} := eq_refl.
End M.

Module F (A:T).
  Definition bad := TypeNeqSmallType.paradox _ A.e.
End F.

Module X := F M.
Check X.bad.
(* X.bad
     : False
*)
Print Assumptions X.bad.
(* Closed under the global context *)
```
The checker does catch this one (ill-typed term: CantApplyBadType).

#7490 blocks it for coqtop.

### PR DESCRIPTION
**Pull request by SkySkimmer**
When inferring [u <= v+k] I replaced the exception and instead add
[u <= v]. This is trivially sound and it doesn't seem possible to have
the one without the other (except specially for [Set <= v+k] which was
already handled).

I don't know an example where this used to fail and now succeeds (the
point was to remove an anomaly, but the example
~~~coq
Module Type SG. Definition DG := Type. End SG.
Module MG : SG. Definition DG := Type : Type. Fail End MG.
~~~
now fails with universe inconsistency.

Fix #7695 (soundness bug!).

### CODE REVIEW

**Review by mattam82 — APPROVED** [review]
Good, I was about to say we could just be incomplete in the treatment of i <= j+k constraints :) The patch looks good to me. From what I understand, the invariant for the inferred case holds and can be checked with this implementation. The main point is that we can check max() <= max() constraints modulo the u <= v+k ~> u <= v approximation.

File: checker/subtyping.ml [hunk 1]
```diff
@@ -134,40 +133,7 @@ let check_inductive  env mp1 l info1 mib2 spec2 subst1 subst2=
...
-    let (ctx2,s2) = dest_arity env t2 in
-    let s1,s2 =
-      match s1, s2 with
-      | Type _, Type _ -> (* shortcut here *) Prop Null, Prop Null
-      | (Prop _, Type _) | (Type _,Prop _) -> error ()
-      | _ -> (s1, s2) in
-    check_conv conv_leq env
-      (mkArity (ctx1,s1)) (mkArity (ctx2,s2))
-  in
+  let check_inductive_type t1 t2 = check_conv conv_leq env t1 t2 in
```
Comment (gares) [inline checker/subtyping.ml]: This is not a very deep comment, but this function is used only once and also has a confusing name (you are checking the type of an inductive type). I suggest to inline this.

File: checker/subtyping.ml [hunk 2]
```diff
@@ -253,52 +219,7 @@ let check_inductive  env mp1 l info1 mib2 spec2 subst1 subst2=
...
-                 (the user has to use an explicit type in the interface *)
-                error ()
-          with UserError _ (* "not an arity" *) ->
-            error () end
-        | _ -> t1,t2
-      else
-        (t1,t2) in
-    check_conv conv_leq env t1 t2
-  in
+  let check_type env t1 t2 = check_conv conv_leq env t1 t2 in
```
Comment (gares) [inline checker/subtyping.ml]: ditto

File: checker/subtyping.ml (hunk 1 above)
Comment (gares) [reply checker/subtyping.ml]: Also looks like you are mostly reverting e4dfd24e4a4d4251c79694d82b1aa860657ee6dc (before that the function was not there).

**Comment by mattam82** [comment]
Indeed @gares comment's would make the code more regular. You can also remove the misleading comment about "not checking sorts of polymorphic types" I believe.

**Review by ppedrot — COMMENTED** [review]
I don't feel qualified for this particular kernel patch.

**Comment by Zimmi48** [comment]
It looks like this is making VST fail with a universe inconsistency. Unless, there is a way to adapt the fix to still be sound and have VST compile as-is, this will require an overlay. Since this fixes a critical bug, the overlay won't be a reason not to backport.

**Comment by Zimmi48** [comment]
@SkySkimmer Do you intend to finish this or would you like some help?

**Comment by SkySkimmer** [comment]
>@SkySkimmer Do you intend to finish this or would you like some help?

I don't have time to look at the VST inconsistency. If you do feel free.

**Comment by Zimmi48** [comment]
I have been working on reducing the problem, I'll submit the example soon.

**Comment by Zimmi48** [comment]
OK, so it's a very very weird bug:

File `abcdefghijklmno.v` contains:

```
Definition pred (A:Type) := A -> Prop.
```

File `bug1.v` contains:

```
Require Import abcdefghijklmno.

Module Type bar.
  Parameter T : Type.
  Definition foo := pred T.
End bar.

Module foobar : bar.
  Parameter F : Type -> Type.
  Definition T := F Type.
  Definition foo := T -> Prop.
End foobar.
```

This gives a universe inconsistency.
But try to replace `Prop` with `Type` or inline the definition of `pred` or include the module `abcdefghijklmno` or shorten the name of the `abcdefghijklmno.v` file and you will get no universe inconsistency!

**Comment by Zimmi48** [comment]
I hope that someone can take over from here because it's well out my expertise and I've run out of time to give to this bug.

**Comment by SkySkimmer** [comment]
Looks like `enforce_leq` has been broken since e759333a8b5c11247c4cc134fdde8c1bd85a6e17.
`acc` gets dropped in the recursion https://github.com/coq/coq/blob/6715e6801c1d285a12eeca55dd8b831d7efb8c0d/kernel/univ.ml#L677-L682
(this causes the instability wrt the file name (combined with the order of levels in a Universe.t being determined by the hash), ie with it fixed you always get an inconsistency)

**Comment by SkySkimmer** [comment]
See #7869 (as I said it only makes the error happen consistently, so not a dependency for this PR).

**Comment by Zimmi48** [comment]
> See #7869 (as I said it only makes the error happen consistently, so not a dependency for this PR).

Would you mind rebasing on this PR though to check whether there is still an incompatibility in VST?

**Comment by SkySkimmer** [comment]
rebased

Gaëtan Gilbert

**Comment by Zimmi48** [comment]
Thanks but you didn't rebase on top of the other PR.

**Comment by SkySkimmer** [comment]
With #7869 VST is still broken https://gitlab.com/SkySkimmer/coq/-/jobs/76549380

If we use a smarter `enforce_leq` it becomes fixed https://gitlab.com/SkySkimmer/coq/-/jobs/76581515
See 2c198ca6f18eea1eea38babb77728a6895c07acd
When inferring `max(u_i) <= max(v_j)` (`<-> forall i, exists j, u_i <= v_j`) we look for a solution which is valid in the current ugraph.

**Comment by Zimmi48** [comment]
Thanks! Maybe @mattam82 @maximedenes @ppedrot can comment on the opportunity to add this commit? It would be better if we could avoid breaking VST in a patch-level release of course... but for a critical bug fix there can be an exception as I said above.

**Comment by SkySkimmer** [comment]
Added 2c198ca and parents to this PR. This should remove the error in VST.

**Review by mattam82 — APPROVED** [review]
As discussed on the commit, I agree with the change. Nice beautifying of the code BTW :)

**Comment by mattam82** [comment]
Shall I merge this one and close https://github.com/coq/coq/pull/7869 now?

**Comment by mattam82** [comment]
I'm a bit worried of the fiat-crypto failure, is it unrelated?

**Comment by SkySkimmer** [comment]
Let's restart it.

**Comment by mattam82** [comment]
Ok, looks good now.

**Comment by SkySkimmer** [comment]
Bit late but here's a bench (just started) https://ci.inria.fr/coq/job/benchmark-part-of-the-branch/460/console

**Comment by SkySkimmer** [comment]
```

┌──────────────────────────┬─────────────────────────┬───────────────────────────────────────┬───────────────────────────────────────┬─────────────────────────┬─────────────────┐
│                          │      user time [s]      │              CPU cycles               │           CPU instructions            │  max resident mem [KB]  │   mem faults    │
│                          │                         │                                       │                                       │                         │                 │
│             package_name │     NEW     OLD PDIFF   │            NEW            OLD PDIFF   │            NEW            OLD PDIFF   │     NEW     OLD PDIFF   │ NEW OLD PDIFF   │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│      coq-formal-topology │   37.74   37.98 -0.63 % │   101551370480   101804958613 -0.25 % │   126399621928   126566127752 -0.13 % │  479176  478952 +0.05 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│   coq-mathcomp-odd-order │ 1409.01 1417.44 -0.59 % │  3914883163675  3938517220257 -0.60 % │  6639913839137  6637955673817 +0.03 % │ 1370244 1377892 -0.56 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                coq-flocq │  260.32  261.85 -0.58 % │   721914630278   727224464941 -0.73 % │   897335901277   897714722515 -0.04 % │ 1244568 1244572 -0.00 % │   3   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                  coq-vst │ 3739.88 3759.71 -0.53 % │ 10397025362703 10451968209461 -0.53 % │ 13666181332899 13666685700049 -0.00 % │ 2229424 2228648 +0.03 % │   0   1 -100.00 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                 coq-corn │ 1540.51 1548.60 -0.52 % │  4262670146454  4284642991436 -0.51 % │  6321444673231  6321566343534 -0.00 % │  852912  852812 +0.01 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                 coq-hott │  309.13  310.71 -0.51 % │   859478324948   862309973186 -0.33 % │  1274362934755  1274792069029 -0.03 % │  628192  586756 +7.06 % │  27  26 +3.85 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│       coq-mathcomp-field │  447.92  449.34 -0.32 % │  1243028564283  1246460815919 -0.28 % │  2026499853515  2026602504061 -0.01 % │  803484  803464 +0.00 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│    coq-mathcomp-fingroup │   65.16   65.35 -0.29 % │   178752001029   179371251849 -0.35 % │   238620907685   238826440790 -0.09 % │  586268  586532 -0.05 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│              coq-bignums │   79.13   79.35 -0.28 % │   218795402256   219014888978 -0.10 % │   282909304913   283135597613 -0.08 % │  519284  519080 +0.04 % │   1   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│   coq-mathcomp-character │  272.94  273.61 -0.24 % │   756254323353   759293346997 -0.40 % │  1080238117926  1080339639973 -0.01 % │ 1127892 1127748 +0.01 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│    coq-mathcomp-solvable │  200.86  201.16 -0.15 % │   555632735345   557237966998 -0.29 % │   767133753256   767307350655 -0.02 % │  842432  822460 +2.43 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│               coq-geocoq │ 2799.84 2803.76 -0.14 % │  7762540631450  7771121004118 -0.11 % │ 12490527506439 12489987809493 +0.00 % │ 1387748 1387652 +0.01 % │   5   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│   coq-mathcomp-ssreflect │   45.66   45.72 -0.13 % │   124416944576   124420549940 -0.00 % │   145973551950   146160725012 -0.13 % │  532832  533012 -0.03 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│            coq-fiat-core │  111.43  111.52 -0.08 % │   310249056616   309687550542 +0.18 % │   378554608222   378733744502 -0.05 % │  495468  495152 +0.06 % │  57   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│     coq-mathcomp-algebra │  184.58  184.60 -0.01 % │   509685138448   510056813797 -0.07 % │   686805021044   687131146486 -0.05 % │  640036  640044 -0.00 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│ coq-mathcomp-real-closed │  168.06  168.00 +0.04 % │   465774611134   466275062347 -0.11 % │   683895807171   684363444963 -0.07 % │  826980  825440 +0.19 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│           coq-coquelicot │   81.93   81.87 +0.07 % │   226180761585   226097993160 +0.04 % │   268010016489   268145015704 -0.05 % │  709512  709808 -0.04 % │   9  83 -89.16 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│              coq-unimath │ 1280.08 1279.13 +0.07 % │  3539541661007  3538367292442 +0.03 % │  6025060592102  6025140013281 -0.00 % │ 1054300 1054316 -0.00 % │   2   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                coq-color │  715.17  713.40 +0.25 % │  1974517197952  1972378915210 +0.11 % │  2347162479360  2346953772481 +0.01 % │ 1387656 1387284 +0.03 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│               coq-sf-vfa │   25.89   25.82 +0.27 % │    70199586703    70032021497 +0.24 % │    90360174044    90539235748 -0.20 % │  531996  532196 -0.04 % │  24  24 +0.00 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│         coq-math-classes │  226.59  225.77 +0.36 % │   622523640366   619146032672 +0.55 % │   770058534858   770256480290 -0.03 % │  532624  533032 -0.08 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│         coq-fiat-parsers │  702.22  699.36 +0.41 % │  1944548833007  1933970643370 +0.55 % │  2974571228929  2974975229603 -0.01 % │ 3367392 3366280 +0.03 % │   0 354 -100.00 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│               coq-sf-plf │   52.19   51.92 +0.52 % │   142722495113   142688656002 +0.02 % │   178610932341   178707450538 -0.05 % │  515416  515356 +0.01 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│             coq-compcert │  870.24  863.64 +0.76 % │  2413196240600  2393217067114 +0.83 % │  3457944503319  3458385142601 -0.01 % │ 1348540 1348328 +0.02 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                coq-sf-lf │   20.05   19.86 +0.96 % │    53511667840    53620079271 -0.20 % │    69044566052    69242227539 -0.29 % │  416596  416584 +0.00 % │   3   0  +nan % │
└──────────────────────────┴─────────────────────────┴───────────────────────────────────────┴───────────────────────────────────────┴─────────────────────────┴─────────────────┘


```

**Comment by mattam82** [comment]
Good, thanks!

**Comment by Zimmi48** [comment]
@SkySkimmer This is not trivial to backport. Would you mind taking care of it and opening a PR? Thanks!

**Pull request merged**